tsdb/importer/import.go

```diff
@@ -0,0 +1,310 @@
+							return blockPaths, err
+						}
+						blockPaths = append(blockPaths, path)
+					}
+					currentBlock = []*tsdb.MetricSample{}
+					if haveSamplesInsideBlockTimeRange {
+						haveSamplesInsideBlockTimeRange = false
+					}
+				}
+			}
+			currentBlock = append(currentBlock, sample)
+			continue
+		}
+		// This will only be triggered in the case where we have some
+		// samples from before the DB mint, that also overlap with the
+		// DB blocks times.
+		if haveSamplesBeforeDbStart {
+			parts := partitionBlocks(currentBlock, tsdb.DefaultBlockDuration)
+			for _, part := range parts {
+				path, err := tsdb.CreateBlock(part, dbPath, logger)
+				if err != nil {
+					return blockPaths, err
+				}
+				blockPaths = append(blockPaths, path)
+			}
+			currentBlock = []*tsdb.MetricSample{}
+			// Once this is set to false in here, it will never be triggered again,
+			// as we have already crossed into the DB's time limits now.
+			haveSamplesBeforeDbStart = false
+		}
+		if blockTimes[blockIdx].start <= sample.TimestampMs && sample.TimestampMs <= blockTimes[blockIdx].end {
+			currentBlock = append(currentBlock, sample)
+			haveSamplesInsideBlockTimeRange = true
+		} else {
+			if len(currentBlock) > 0 {
+				// Push to disk, no need to partition.
+				path, err := tsdb.CreateBlock(currentBlock, dbPath, logger)
+				if err != nil {
+					return blockPaths, err
+				}
+				blockPaths = append(blockPaths, path)
+				blockIdx++
+			}
+			// Increment blockIdx until we reach a block which can accept our sample.
+			inWindow := false
+			for !inWindow && blockIdx < len(blockTimes) {
+				if blockTimes[blockIdx].start <= sample.TimestampMs && sample.TimestampMs <= blockTimes[blockIdx].end {
+					inWindow = true
+				} else {
+					blockIdx++
+				}
+			}
+			currentBlock = []*tsdb.MetricSample{sample}
+			haveSamplesInsideBlockTimeRange = true
+		}
+	}
+	// If there are any blocks that haven't yet been pushed to disk.
+	if len(currentBlock) > 0 {
+		parts := partitionBlocks(currentBlock, tsdb.DefaultBlockDuration)
+		for _, part := range parts {
+			path, err := tsdb.CreateBlock(part, dbPath, logger)
+			if err != nil {
+				return blockPaths, err
+			}
+			blockPaths = append(blockPaths, path)
+		}
+	}
+	return blockPaths, nil
+}
+
+// partitionBlocks partitions samples into blocks of the specified duration.
+func partitionBlocks(samples []*tsdb.MetricSample, duration int64) [][]*tsdb.MetricSample {
+	var blocks [][]*tsdb.MetricSample
+	var currentBlock []*tsdb.MetricSample
+	var startTime time.Time
+	var currentTime time.Time
+	for _, currentSample := range samples {
+		secs, nsecs := int64(currentSample.TimestampMs/1000), int64((currentSample.TimestampMs%1000)*1000)
+		currentTime = time.Unix(secs, nsecs)
+		if startTime.IsZero() {
+			startTime = currentTime
+			currentBlock = append(currentBlock, currentSample)
+			continue
+		}
+		timeDelta := currentTime.Sub(startTime)
+		if int64(timeDelta.Seconds()*1000) >= duration {
```

Comment:
This doesn't ensure alignment with the existing blocks

Reply:
Well, this function basically just partitions the samples into bins of certain durations; I perform the actual alignment in the `pushMetrics` function itself.

Reply:
I'm not seeing that, you appear to create blocks directly from the output of this function